drivers/serial/uart_nrfx_uarte.c

```diff
@@ -1040,69 +1123,66 @@ static int uarte_nrfx_poll_in(const struct device *dev, unsigned char *c)
  */
 static void uarte_nrfx_poll_out(const struct device *dev, unsigned char c)
 {
-	NRF_UARTE_Type *uarte = get_uarte_instance(dev);
 	struct uarte_nrfx_data *data = get_dev_data(dev);
-	atomic_t *lock;
-
-#ifdef CONFIG_UART_ASYNC_API
+	bool isr_mode = k_is_in_isr() || k_is_pre_kernel();
+	NRF_UARTE_Type *uarte = get_uarte_instance(dev);
+	int key;
 
-	if (data->async) {
-		while (data->async->tx_buf) {
-			/* If there is ongoing transmission, and we are in
-			 * isr, then call uarte interrupt routine, otherwise
-			 * busy wait until transmission is finished.
-			 */
-			if (k_is_in_isr()) {
+	if (isr_mode) {
+		while (1) {
+#if CONFIG_UART_ASYNC_API
+			if (data->async) {
 				uarte_nrfx_isr_async(dev);
 			}
-		}
-		/* Use tx_size as lock, this way uarte_nrfx_tx will
-		 * return -EBUSY during poll_out.
-		 */
-		lock = &data->async->tx_size;
-	} else
 #endif
-		lock = &data->poll_out_lock;
 
-	if (!k_is_in_isr()) {
-		uint8_t safety_cnt = 100;
+			if (data->uart_config.flow_ctrl ==
+			    UART_CFG_FLOW_CTRL_NONE) {
+				bool res;
 
-		while (atomic_cas((atomic_t *) lock,
-				(atomic_val_t) 0,
-				(atomic_val_t) 1) == false) {
-			/* k_sleep allows other threads to execute and finish
-			 * their transactions.
-			 */
-			k_msleep(1);
-			if (--safety_cnt == 0) {
+				NRFX_WAIT_FOR(is_tx_ready(dev),
+					      POLL_OUT_TIMEOUT_US, 1, res);
```

Comment:
Why do we need this timeout? Is this possible that the event won't eventually appear in a scenario with the flow control disabled?

Reply:
will remove